libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -596,15 +603,17 @@ instance HasSpec fn a => Semigroup (Specification fn a) where
   ErrorSpec e <> ErrorSpec e' = ErrorSpec (e ++ "" : (replicate 20 '-') : "" : e')
   ErrorSpec e <> _ = ErrorSpec e
   _ <> ErrorSpec e = ErrorSpec e
-  MemberSpec as <> MemberSpec as' =
-    explainSpec ["Intersecting: ", "  MemberSpec " ++ show as, "  MemberSpec " ++ show as'] $
-      MemberSpec $
-        intersect as as'
-  MemberSpec as <> TypeSpec s cant =
-    MemberSpec $
-      filter
-        (flip (conformsTo @fn) s)
-        (filter (`notElem` cant) as)
+  MemberSpec as <> MemberSpec as' = case intersect as as' of
+    [] -> nullMemberMember as as'
+    other -> MemberSpec other
+  rhs@(MemberSpec as) <> lhs@(TypeSpec s cant) =
+    case filter (flip (conformsTo @fn) s) (filter (`notElem` cant) as) of
+      x@(_ : _) -> MemberSpec x
+      [] ->
+        ErrorSpec
+          [ "\nNull intersection of MemberSpecs and TypeSpec"
+          , "  " ++ show rhs ++ " <> " ++ show lhs
+          ]
```

Comment:
This is being addressed in #4317

Reply:
Not the latter case, sorry. That one might make sense to include here. However, the extra explanation in `MemberSpec` shouldn't be necessary now because we provide much more contexts in the failing cases now.

Reply:
OK, I will leave this here for now. After PR #4317 is merged into master, I will rebase and set things straight.